accounts/abi/bind/backends/simulated_test.go

```diff
@@ -397,25 +399,120 @@ func TestSimulatedBackend_TransactionByHash(t *testing.T) {
+
+	key, _ := crypto.GenerateKey()
+	addr := crypto.PubkeyToAddress(key.PublicKey)
+	opts := bind.NewKeyedTransactor(key)
+
+	sim := simTestBackend(addr)
 	defer sim.Close()
-	bgCtx := context.Background()
-	testAddr := crypto.PubkeyToAddress(testKey.PublicKey)
 
-	gas, err := sim.EstimateGas(bgCtx, klaytn.CallMsg{
-		From:  testAddr,
-		To:    &testAddr,
-		Value: big.NewInt(1000),
-		Data:  []byte{},
-	})
-	if err != nil {
-		t.Errorf("could not estimate gas: %v", err)
-	}
+	parsed, _ := abi.JSON(strings.NewReader(contractAbi))
+	contractAddr, _, _, _ := bind.DeployContract(opts, parsed, common.FromHex(contractBin), sim)
+	sim.Commit()
 
-	if gas != params.TxGas {
-		t.Errorf("expected 21000 gas cost for a transaction got %v", gas)
+	cases := []struct {
+		name        string
+		message     klaytn.CallMsg
+		expect      uint64
+		expectError error
+		expectData  interface{}
+	}{
+		{"plain transfer(valid)", klaytn.CallMsg{
+			From:     addr,
+			To:       &addr,
+			Gas:      0,
+			GasPrice: big.NewInt(0),
+			Value:    big.NewInt(1),
+			Data:     nil,
+		}, params.TxGas, nil, nil},
+
+		{"plain transfer(invalid)", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      0,
+			GasPrice: big.NewInt(0),
+			Value:    big.NewInt(1),
+			Data:     nil,
+		}, 0, errors.New("evm: execution reverted"), nil},
+
+		{"Revert", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      0,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("d8b98391"),
+		}, 0, errors.New("execution reverted: revert reason"), "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000d72657665727420726561736f6e00000000000000000000000000000000000000"},
+
+		{"PureRevert", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      0,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("aa8b1d30"),
+		}, 0, errors.New("evm: execution reverted"), nil},
+
+		{"OOG", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      100000,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("50f6fe34"),
+		}, 0, errors.New("gas required exceeds allowance (100000)"), nil},
+
+		{"Assert", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      100000,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("b9b046f9"),
+		}, 0, errors.New("VM error occurs while running smart contract"), nil},
+
+		{"Valid", klaytn.CallMsg{
+			From:     addr,
+			To:       &contractAddr,
+			Gas:      100000,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("e09fface"),
+		}, 21483, nil, nil},
+	}
+	for _, c := range cases {
+		got, err := sim.EstimateGas(context.Background(), c.message)
+		if c.expectError != nil {
```

Comment:
confused a little bit. Is it matter if the expectError is nil or not?

if the expectedError is of the same type as the value returned by the function, I think the following code is enough
```go
if c.expectError != err {
    t.Fatalf("Expect error, want %v, got %v", c.expectError, err)
}
if c.expectData != nil {
    if err, ok := err.(*revertError); !ok {
        t.Fatalf("Expect revert error, got %T", err)
    } else if !reflect.DeepEqual(err.ErrorData(), c.expectData) {
        t.Fatalf("Error data mismatch, want %v, got %v", c.expectData, err.ErrorData())
    }
}
```

Reply:
`c.expectError != err` doesn't work. It has to check `err.Error()`. But, if before calling Error(), it needs to check err is nil or not.

Reply:
I got it, thanks